src/cpp/rtps/history/History.cpp

```diff
@@ -100,6 +100,76 @@ void History::do_release_cache(
     change_pool_->release_cache(ch);
 }
 
+History::const_iterator History::find_change(
```

Comment:
Returning an iterator to a Vector. This could be invalidated at any time if the history is reallocated. This iterator should not be used unless protected by a lock. For example, on `remove_change` the lock is released between the find and the removal.

See my comment on that method too, about trying to avoid releasing the lock to avoid context changes.

Reply:
It's exactly the same situation with the stable APIs: `changesBegin` and `changesEnd`.
With the refactor hinted by you the old `remove_change` becomes again atomic.